Max pool

americast opened the pull request:
An implementation of `maxpool`. Here's a sample benchmarking (CPU v/s GPU): https://gist.github.com/americast/b518a7007c5a3db5a9718a7f4e4a3031

Tests were failing due to `getindex is disabled` error. I have made a small change in `src/indexing.jl` as a workaround.

Thanks.

SimonDanisch commented:
Lol, the assertslow is there for a reason, since scalar indexing into a gpuarray is very slow and should not be done! 
You can do `allowslow(true)` to disable that error. Since your benchmarks seem to indicate good speed, I'm guessing the problem happens on the other side of the test code. I'll check.

americast commented:
Ouch! The benchmarks don't include padding. I should have mentioned that. Sorry, my bad. I'll create a separate benchmark with padding.

americast commented on src/pool.jl:
```diff
@@ -0,0 +1,44 @@
+import CUDAnative
```
:+1:

americast commented on src/pool.jl:
```diff
@@ -0,0 +1,44 @@
+    pool = UInt32(pool)
+    stride = UInt32(stride)
+    out = similar(b)
+    out = out[1:(div(Asize[1] - pool, stride) + 1), 1:(div(Asize[2] - pool, stride) + 1), :, :]
```
Thanks, I was unaware of this. It should be `similar(b, outSize...)` perhaps. Also, `outSize` needs to be determined before `similar` is called.

americast commented:
Updated. Thank you @SimonDanisch for PR #111 and commit 1e1104e. 
I've updated the benchmarks at https://gist.github.com/americast/95358d972647adf5c7ebcde7c58db51f.
